resolwe/test/testcases/__init__.py

```diff
@@ -99,66 +55,78 @@ def _pre_setup(self, *args, **kwargs):
+            FLOW_EXECUTOR_SETTINGS['UPLOAD_DIR'],
+            FLOW_EXECUTOR_SETTINGS['RUNTIME_DIR'],
+        )
+        # Directories need to be recreated here in case a previous
+        # TestCase deleted them. Moving this logic into the test runner
+        # and manager infrastructure would not work, because the manager
+        # and listener can't know where the testcase boundaries are,
+        # they just see a series of data objects; deleting too soon
+        # might cause problems for some tests. The runner does not have
+        # any code between tests, so could only remove data at the very
+        # end, by which time it's already too late, since some tests may
+        # deal specifically with the purging functionality and should
+        # start in a clean environment, without the sediment
+        # (e.g. jsonout.txt, stdout.txt) from previous tests.
+        os.makedirs(data_dir, exist_ok=True)
+        os.makedirs(upload_dir, exist_ok=True)
+        os.makedirs(runtime_dir, exist_ok=True)
+
+        # Override container name prefix setting
+        flow_executor_settings['CONTAINER_NAME_PREFIX'] = '{}_{}_{}'.format(
+            getattr(settings, 'FLOW_EXECUTOR', {}).get('CONTAINER_NAME_PREFIX', 'resolwe'),
+            # NOTE: This is necessary to avoid container name clashes when tests are run from
+            # different Resolwe code bases on the same system (e.g. on a CI server).
+            get_random_string(length=6),
+            os.path.basename(data_dir)
+        )
+
+        # Override Docker data directory mappings
+        flow_docker_mappings = copy.copy(getattr(settings, 'FLOW_DOCKER_MAPPINGS', []))
+        for mapping in flow_docker_mappings:
+            if mapping['dest'] == '/data':
+                mapping['src'] = os.path.join(data_dir, '{data_id}')
+            elif mapping['dest'] == '/data_all':
+                mapping['src'] = data_dir
+            elif mapping['dest'] == '/upload':
+                mapping['src'] = upload_dir
+
+        self.settings = override_settings(FLOW_EXECUTOR=flow_executor_settings,
+                                          FLOW_DOCKER_MAPPINGS=flow_docker_mappings)
         self.settings.enable()
 
-        self._keep_data = False
+        self._keep_data = settings.FLOW_MANAGER_TEST_KEEP_DATA
 
         self.addCleanup(self._clean_up)
 
     def keep_data(self, mock_purge=True):
-        """Do not delete output files after test."""
-        self._keep_data = True
-
-        if mock_purge:
-            purge_mock_os = mock.patch('resolwe.flow.utils.purge.os', wraps=os).start()
-            purge_mock_os.remove = mock.MagicMock()
-
-            purge_mock_shutil = mock.patch('resolwe.flow.utils.purge.shutil', wraps=shutil).start()
-            purge_mock_shutil.rmtree = mock.MagicMock()
+        """Do not delete output files after tests."""
+        print("\n*ERROR* TestCaseHelpers.keep_data() is deprecated and does not work anymore.")
```

Comment:
Any reason why this message is not part of the exception? This way, the user would not need to scroll up to see the actual deprecation message. This is especially true in tests, where the print output may be buried quite deep.

Reply:
The reasoning was to keep the exception message short enough for the traceback exception message line. I could get rid of the prints altogether and stuff everything into the exception message, or just add a note about the command line options to the exception message. Which would be better?

Reply:
I would extend the exception message, similar to what [@kostko has done if a process doesn't have `@tag_process`](https://github.com/genialis/resolwe/blob/82fc0f38a43c5378f4f1427159bdca4a17d9930d/resolwe/test/testcases/process.py#L348-L355).
Maybe you could also use `self.fail` instead of `raise RuntimeError`?